### PR TITLE
Text editor, move to this repo

### PR DESCRIPTION
**Pull request by georgiana-ecobici**
Change-Id: Id9a163cbb5f60e358b2155a801de0f4e45246289
Signed-off-by: Georgiana-Elena ECOBICI <georgiana-elena.ecobici@thalesgroup.com>

### CODE REVIEW

**Review by minhtutonthat — CHANGES_REQUESTED** [review]
First remarks about the structure of plugins:
* The project org.polarsys.capella.scenario.editor.dsl should not contain other projects. All projects should be put at the same level in the plugins folder.
* All test projects should be put in tests folder
* TestDsl is a folder that contains samples, right? It should be placed in the samples folder

**Comment by minhtutonthat** [comment]
For this version, I've done a quick test for the synchronization between the diagram and the editor. The inital diagram contains only instance roles and sequence messages:


The synchronization from diagram to editor (via the Refresh button) works well. When the content of the diagram changes, the modifications are well reflected in the editor. However, it does not work the opposite way (from editor to diagram). When I change the order of instance roles and sequence messages, the diagram does not show my changes in the editor and it adds new sequence messages to the diagram.

**Comment by minhtutonthat** [comment]
About the language and generated code
* Could you change the name of the ecore from "Dsl" to something more meaningful (e.g. TextualScenario) 
* Some generated codes do not have the correct copyright. Maybe you should replace the Copyright text field in the Dsl.genmodel by our copyright.

File: plugins/org.polarsys.capella.scenario.editor/src/org/polarsys/capella/scenario/editor/EmbeddedEditorInstance.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  }
+
+  public EmbeddedEditorInstance getInstance() {
+    if (eINSTANCE == null) {
+      eINSTANCE = new EmbeddedEditorInstance();
+    }
+    return eINSTANCE;
+  }
+
+  public static EmbeddedEditor geteEditor() {
```
Comment (minhtutonthat) [inline plugins/org.polarsys.capella.scenario.editor/src/org/polarsys/capella/scenario/editor/EmbeddedEditorInstance.java]: Could you change the variable name to embeddedEditor and change method names accordingly?

File: plugins/org.polarsys.capella.scenario.editor/src/org/polarsys/capella/scenario/editor/helper/EmbeddedEditorInstanceHelper.java [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+import org.polarsys.capella.core.sirius.analysis.OAServices;
+import org.polarsys.capella.scenario.editor.EmbeddedEditorInstance;
+
+public class EmbeddedEditorInstanceHelper {
+  public static EList<InstanceRole> getAvailableInstanceRoles() {
+    Scenario currentScenario = EmbeddedEditorInstance.getAssociatedScenarioDiagram();
+    return currentScenario.getOwnedInstanceRoles();
+  }
+
+  public static String getName(EObject element) {
```
Comment (minhtutonthat) [inline plugins/org.polarsys.capella.scenario.editor/src/org/polarsys/capella/scenario/editor/helper/EmbeddedEditorInstanceHelper.java]: Use org.polarsys.capella.core.model.helpers.CapellaElementExt.getName(EObject) instead.

File: plugins/org.polarsys.capella.scenario.editor/src/org/polarsys/capella/scenario/editor/helper/EmbeddedEditorInstanceHelper.java [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+      if (ScenarioExt.isFunctionalScenario(currentScenario)) {
+        return keyword.equals(DslConstants.ACTIVITY);
+      } else {
+        return keyword.equals(DslConstants.ENTITY) || keyword.equals(DslConstants.ACTOR)
+            || keyword.equals(DslConstants.ROLE);
+      }
+    }
+    // IS and ES
+    if (currentScenario.getKind() == ScenarioKind.INTERFACE || currentScenario.getKind() == ScenarioKind.DATA_FLOW) {
+      if (blockArchitecture instanceof SystemAnalysis) {
```
Comment (minhtutonthat) [inline plugins/org.polarsys.capella.scenario.editor/src/org/polarsys/capella/scenario/editor/helper/EmbeddedEditorInstanceHelper.java]: Could we merge these 2 if?

File: plugins/org.polarsys.capella.scenario.editor/src/org/polarsys/capella/scenario/editor/helper/EmbeddedEditorInstanceHelper.java [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+
+  /**
+   * returns the list of available elements that could be inserted for the given keyword (which can be actor, component,
+   * function, entity, role, activity, configuration_item)
+   * 
+   * @param keyword
+   * @return boolean
+   *
+   */
+  public static Collection<? extends EObject> getAvailableElements(String keyword) {
```
Comment (minhtutonthat) [inline plugins/org.polarsys.capella.scenario.editor/src/org/polarsys/capella/scenario/editor/helper/EmbeddedEditorInstanceHelper.java:225]: Could we have separate methods such as getAvailableAbstractFunction, getAvailableRoles, getAvailableActors, etc. to replace the method getAvailableElements which groups everything in the same package.

File: plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextEditorCommands.java [hunk 5]
```diff
@@ -0,0 +1,540 @@
...
+import org.polarsys.capella.scenario.editor.dslscenario.dsl.Participant;
+import org.polarsys.capella.scenario.editor.dslscenario.dsl.ParticipantDeactivation;
+import org.polarsys.capella.scenario.editor.dslscenario.dsl.impl.DslFactoryImpl;
+import org.polarsys.capella.scenario.editor.dslscenario.dsl.impl.ModelImpl;
+import org.polarsys.capella.scenario.editor.dslscenario.ui.provider.DslscenarioProvider;
+import org.polarsys.capella.scenario.editor.embeddededitor.views.EmbeddedEditorView;
+import org.polarsys.capella.scenario.editor.helper.DslConstants;
+import org.polarsys.capella.scenario.editor.helper.EmbeddedEditorInstanceHelper;
+
+public class XtextEditorCommands {
```
Comment (minhtutonthat) [inline plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextEditorCommands.java]: Could you separate this class into 2 classes, one that transform text to diagram and the other from diagram to text?

File: plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextEditorCommands.java [hunk 6]
```diff
@@ -0,0 +1,540 @@
...
+      // Project project;
+      BlockArchitecture blockArchitecture = BlockArchitectureExt.getRootBlockArchitecture(scenario);
+
+      doEditingOnParticipants(scenario, blockArchitecture, participants);
+
+      doEditingOnMessages(scenario, blockArchitecture, messages);
+    }
+  }
+
+  private static void doEditingOnParticipants(Scenario scenario, BlockArchitecture blockArchitecture,
```
Comment (minhtutonthat) [inline plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextEditorCommands.java]: To reimplement methods such as doEditingOnParticipants and doEditingOnMessages. Now they just simply create new elements and add them to the list. They should:
* Add new element if not exist
* Delete existing element if removed
* Reorder existing element if its position has changed

Like this we can maintain the elements that have existed in the diagram and update it with the modifications in the textual editor.

File: plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextEditorCommands.java [hunk 7]
```diff
@@ -0,0 +1,540 @@
...
+          EObject messageFromXtext = iterator.next();
+
+          if (messageFromXtext instanceof org.polarsys.capella.scenario.editor.dslscenario.dsl.SequenceMessage) {
+
+            org.polarsys.capella.scenario.editor.dslscenario.dsl.SequenceMessage seqMessage = (org.polarsys.capella.scenario.editor.dslscenario.dsl.SequenceMessage) messageFromXtext;
+            InstanceRole source = EmbeddedEditorInstanceHelper.getInstanceRole(seqMessage.getSource());
+            InstanceRole target = EmbeddedEditorInstanceHelper.getInstanceRole(seqMessage.getTarget());
+            if (source != null && target != null) {
+              // create Capella SequenceMessage
+              sequenceMessage = InteractionFactory.eINSTANCE.createSequenceMessage();
```
Comment (minhtutonthat) [inline plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextEditorCommands.java]: Could you extract these codes to helper methods?

File: plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextEditorCommands.java [hunk 8]
```diff
@@ -0,0 +1,540 @@
...
+    return participantDeactivationMsg;
+  }
+
+  private static EObject copyMessageFromMsgEnd(Object object, DslFactory factory) {
+    EObject seqMessage;
+    MessageEnd end = (MessageEnd) object;
+    SequenceMessage sequenceMessage = end.getMessage();
+
+    switch (sequenceMessage.getKind()) {
+    case ASYNCHRONOUS_CALL:
```
Comment (minhtutonthat) [inline plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextEditorCommands.java]: There are duplicated block in the switch. Could you merge them?

**Review by minhtutonthat — CHANGES_REQUESTED** [review]

File: plugins/org.polarsys.capella.scenario.editor/src/org/polarsys/capella/scenario/editor/EmbeddedEditorInstance.java (hunk 1 above)
Comment (georgiana-ecobici) [reply plugins/org.polarsys.capella.scenario.editor/src/org/polarsys/capella/scenario/editor/EmbeddedEditorInstance.java]: Done, all comments to this point.

File: plugins/org.polarsys.capella.scenario.editor/src/org/polarsys/capella/scenario/editor/helper/EmbeddedEditorInstanceHelper.java (hunk 2 above)
Comment (georgiana-ecobici) [reply plugins/org.polarsys.capella.scenario.editor/src/org/polarsys/capella/scenario/editor/helper/EmbeddedEditorInstanceHelper.java]: Done

File: plugins/org.polarsys.capella.scenario.editor/src/org/polarsys/capella/scenario/editor/helper/EmbeddedEditorInstanceHelper.java (hunk 3 above)
Comment (georgiana-ecobici) [reply plugins/org.polarsys.capella.scenario.editor/src/org/polarsys/capella/scenario/editor/helper/EmbeddedEditorInstanceHelper.java]: Done

File: plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextEditorCommands.java (hunk 6 above)
Comment (georgiana-ecobici) [reply plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextEditorCommands.java]: The sync is incomplete, currently we only create new elements, task is in progress.

File: plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextEditorCommands.java (hunk 7 above)
Comment (georgiana-ecobici) [reply plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextEditorCommands.java]: OK.

File: plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextEditorCommands.java (hunk 8 above)
Comment (georgiana-ecobici) [reply plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextEditorCommands.java]: OK.

File: plugins/org.polarsys.capella.scenario.editor/src/org/polarsys/capella/scenario/editor/helper/EmbeddedEditorInstanceHelper.java (hunk 4 above)
Comment (georgiana-ecobici) [reply plugins/org.polarsys.capella.scenario.editor/src/org/polarsys/capella/scenario/editor/helper/EmbeddedEditorInstanceHelper.java:225]: Done

**Review by minhtutonthat — APPROVED** [review]

**Pull request merged**